fix(forms): incorrectly keeping track of ngModel with ngFor inside a form

crisbeto opened the pull request:
When an `NgModel` is created within a `form`, it receives an `NgControl` based on its `name`, but the control doesn't get swapped out if the name changes. This can lead to problems if the `NgModel` is part of an `ngFor`, because the name can change based on its position in the list and a new control can be defined with the same name, leading us to having multiple directives pointing to the same control. For example, if we start off with a list like :

```
[0, 1, 2]; -> [NgModel(0), NgModel(1), NgModel(2)]
```

Then we remove the second item:

```
[0, 2]; -> [NgModel(0), NgModel(2)]
```

And finally, if we decide to add an item to the end of the list, we'll already have a control for
index 2, causing the list to look like:

```
[0, 2, 3]; -> [NgModel(0), NgModel(2), NgModel(2)]
```

These changes fix the issue by removing the old control when the `name` of the directive changes.

Fixes #38465.
Fixes #37920.

crisbeto commented on packages/forms/src/directives/ng_form.ts:
```diff
@@ -104,7 +104,7 @@ export class NgForm extends ControlContainer implements Form, AfterViewInit {
   public readonly submitted: boolean = false;
 
-  private _directives: NgModel[] = [];
+  private _directives = new Set<NgModel>();
```
`Set` requires a polyfill, but I figured that it's ok to use it here since `core` uses it in a couple of places.

AndrewKushnir replied:
Just to confirm: this is an optimization, which is not directly related to that change? (It's a great optimization, just want to check if I'm missing something)

crisbeto replied:
Apart from it being an optimization, I did it to be safe that we don't accidentally register the same directive twice.

AndrewKushnir replied:
👍

crisbeto commented on packages/forms/src/directives/ng_model.ts:
```diff
@@ -219,7 +219,20 @@ export class NgModel extends NgControl implements OnChanges, OnDestroy {
+          // is inside a promise and it won't run immediately. We work around it by giving it an
+          // object with the same shape instead.
+          const oldName = changes['name'].previousValue;
+          this.formDirective.removeControl({name: oldName, path: this._getPath(oldName)});
```
This part feels a little hacky, but the alternative would've been to add a public API to the `Form` interface.

AndrewKushnir replied:
I was thinking if we can do it a bit safer and retrieve an instance of this directive from parent first and then remove it:
```
  const oldName = changes['name'].previousValue;
  const oldControl = this.formDirective.getControl({path: this._getPath(oldName)});
  this.formDirective.removeControl(oldControl);
```
The `getControl` is pretty simple and it's a "read-only" mode, when `removeControl` is a bit more tricky (since it performs more complex action) and may rely on other control fields in the future.

crisbeto replied:
I was considering doing something similar, but `removeControl` works specifically on `NgModel`, not a `FormControl` which is returned from `getControl`. I could update the signature of `removeControl` to accept both, but it would mean changing the public `Form` interface.

AndrewKushnir replied:
Yeah, ok, thanks for additional info!

AndrewKushnir commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -259,6 +259,61 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
+           @Component({
+             template: `
+              <form>
+                <div *ngFor="let item of items; index as i">
+                  <input [(ngModel)]="item.value" name="name-{{i}}">
```
I'd propose to add one more test with the `ngModelGroup` directive used as a parent, something like:

```
<ng-container ngModelGroup="group">
  <div *ngFor="let item of items; index as i">
    <input [(ngModel)]="item.value" name="name-{{i}}">
  </div>
</ng-container>
```

crisbeto replied:
Added, but I'm not completely sure whether it's what you were looking for. Let me know if I should tweak it.

AndrewKushnir replied:
I've added a quick comment to a new test below.

AndrewKushnir reviewed (commented):
Thanks @crisbeto, the change looks good, just left a few comments. Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:352213989:BASE:352213992:1610843383351:813f64ce) (to see if there is any impact on apps in Google's codebase).

crisbeto committed:
fix(forms): incorrectly keeping track of ngModel with ngFor inside a form

When an `NgModel` is created within a `form`, it receives an `NgControl` based on its `name`, but
the control doesn't get swapped out if the name changes. This can lead to problems if the `NgModel`
is part of an `ngFor`, because the name can change based on its position in the list and a new
control can be defined with the same name, leading us to having multiple directives pointing to
the same control. For example, if we start off with a list like :

```
[0, 1, 2]; -> [NgModel(0), NgModel(1), NgModel(2)]
```

Then we remove the second item:

```
[0, 2]; -> [NgModel(0), NgModel(2)]
```

And finally, if we decide to add an item to the end of the list, we'll already have a control for
index 2, causing the list to look like:

```
[0, 2, 3]; -> [NgModel(0), NgModel(2), NgModel(2)]
```

These changes fix the issue by removing the old control when the `name` of the directive changes.

Fixes #38465.
Fixes #37920.

AndrewKushnir commented:
@crisbeto presubmit indicated a few targets that are affected. I didn't have a chance to perform an investigation, but I'd like to share the error message in case it's helpful:
```
        - Uncaught Error: Uncaught (in promise): TypeError: Cannot read property 'registerControl' of null
        TypeError: Cannot read property 'registerControl' of null
            at packages/forms/src/directives/ng_form.ts?l=191
            at ZoneDelegate.invoke (packages/zone.js/lib/zone.ts?l=1148)
            at FakeAsyncTestZoneSpec.onInvoke (packages/zone.js/lib/zone-spec/fake-async-test.ts?l=674)
            at ZoneDelegate.invoke (packages/zone.js/lib/zone.ts?l=1145)
            at Object.onInvoke (packages/core/src/zone/ng_zone.ts?l=364)
            ...
            at ZoneDelegate.invokeTask (packages/zone.js/lib/zone.ts?l=1182)
```
From the stack trace it's unclear what triggered the error (more investigation is needed).

AndrewKushnir commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -259,6 +259,119 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
+           })
+           class App {
+             private _counter = 0;
+             group = {};
+             items: {value: string}[] = [];
```
I was thinking of the following scenario:
```suggestion
             items: {value: string}[] = [];
             group = {items: this.items};
```
i.e. put items inside the group and test how the fix works with `ngModelGroup` which would act as a parent instead of `NgForm`.

crisbeto commented:
Thank you for the info, if I had to guess, I'd say that [the error comes from this call](https://github.com/angular/angular/blob/master/packages/forms/src/directives/ng_form.ts#L191), although I'm not exactly sure how it can happen.

dylhunn commented:
@crisbeto Now that you can run the presubmit, would you be interested in revisiting this for 14? This would be a nice fix!

crisbeto commented:
Yes, I just need to finish getting some other work submitted first.

crisbeto commented:
@AndrewKushnir @dylhunn I just ran a presubmit and it passed, except for a handful of already failing targets. Maybe the error mentioned above isn't a problem anymore?

AndrewKushnir commented:
> I just ran a presubmit and it passed, except for a handful of already failing targets. Maybe the error mentioned above isn't a problem anymore?

It was likely a target (or a set of targets) from the global run (TGP). Could you please try running it as well?

dylhunn commented:
Choo choo, [all aboard the TAP train](https://test.corp.google.com/OCL:426442739:BASE:426422485:1644001265404:96500991)! 🚂 I'll post back with the global presubmit results.

crisbeto commented:
I ran my own [global presubmit](https://fusion2.corp.google.com/presubmit/tap/426393218/OCL:426393218:BASE:426696388:1644133383264:373ed8c1) as well. It has a bunch of breakages, but they're unrelated and they show up [in other test runs](https://fusion2.corp.google.com/presubmit/tap/425845122/OCL:425845122:BASE:426695903:1644133064512:30d5939b) in other projects as well.

dylhunn commented:
> I ran my own [global presubmit](https://fusion2.corp.google.com/presubmit/tap/426393218/OCL:426393218:BASE:426696388:1644133383264:373ed8c1) as well. It has a bunch of breakages, but they're unrelated and they show up [in other test runs](https://fusion2.corp.google.com/presubmit/tap/425845122/OCL:425845122:BASE:426695903:1644133064512:30d5939b) in other projects as well.

Indeed, my TGP was clean also -- the only breakages pertain to another PR and were fixed elsewhere.

dylhunn reviewed (approved):
reviewed-for: fw-forms

dylhunn commented:
This PR was merged into the repository by commit ebf2fc5224c9debfd839087c9d1d21fb49577d37.
